include/revng/Pipeline/Rank.h

```diff
@@ -16,4 +17,38 @@ class Rank : public DynamicHierarchy<Rank> {
   Rank(llvm::StringRef Name, Rank &Parent) : DynamicHierarchy(Name, Parent) {}
 };
 
+  template<StringTemplateParameter Str, typename LocationType, typename ParentRnk>
```

Comment:
1-2. There's not enough space there 😅 
3. Yeah, I forgot about this!!! Inheriting is insufficient though. `ParentRank` must be a definition. Will make a check though.
4. I hesitated on this one. IMO either way is fine. If it's not `YamlSerializeable` - you'll just get an error when you try to yaml-serialize it. But it's your call, we can restrict it.
5. Yeah, this is a lot better!!!

Reply:
1. `StringTemplateParameter` -> `StringTemplate` or `ConstExprString` or `ConstString`. How does rcc reformats it if it gets too long?
2. Ibidem
3. :+1:
4. Let's be strict, I'm not a fan of APIs that you can only use partially depending on a template argument. Makes sense for super-duper-generic libraries but in regular code it's better to stop the user ASAP. The only point of a Location is to get serialized really.
5. :+1:

Reply:
Sounds good.

Reply:
1-2. I don't like either `StringTemplate` because it can be misunderstood as well as `Const.*` because those are not what this is for (`std::string_view` is perfectly capable of being `constexpr` on its own after all). This trickery is only ever needed if it's used to define a template parameter - hence I wouldn't want to make a `ADT`-namespace name less explicit for the sake of its user.

Reply:
`string_view` is a different concept than `ConstString`. `string_view` doesn't own the data, so there's no overlap in functionality.

Reply:
Actually, if the string is `constexpr`, there's no need to worry about ownership - the compiler guarantees that the lifetime of the literals is as long as the lifetime of the program itself - hence if the string never ever changes - just a view is good enough - hence, in that context, it's the same thing.

Reply:
1. I guess so. It still doesn't sit right with me.

2. I wish we could. There's an extra limitation on the template parameters:

> all base classes and non-static data members are public and non-mutable

for which `std::string_view` doesn't qualify.

Reply:
Types should have their own dignity as standalone creatures.
Therefore type names should represent what the class is, in abstract. It should not represent an exhaustive list of places where it's used.
Sometimes you give names such as `ReturnTypeOfX` for the return type of function `X`, and that's a no-no for me `:)`.

The fact that there's overlap with some other existing tool (`string_view`, `llvm::StringLiteral`...) should not be expressed in the name, maybe in the doc, but not the name.

Reply:
Ended up going with this:
```cpp
template<ConstexprString String,
         detail::YamlizableOrVoid KeyType,
         RankSpecialization ParentRank>
class RankDefinition : public Rank {
// . . .
};
```

What do you think?

Reply:
Actually, no, the keys shouldn't be `void` anywhere but the root. And I did decide to fully separate the root rank from normal ranks for safety sake. As such, we can get away with using `Yamlizable` here: https://github.com/revng/revng/pull/197/commits/a3084ca759dae88f3a295c777f297298a8e04029